src/components/tl/ucp/scatter/scatter_knomial.c

```diff
@@ -0,0 +1,267 @@
+    ucc_tl_ucp_team_t     *team = TASK_TEAM(task);
+    ucc_knomial_pattern_t *p    = &task->scatter_kn.p;
+    ucc_rank_t             vrank, vroot;
+    ucc_status_t           status;
+
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_scatter_kn_start", 0);
+    ucc_tl_ucp_task_reset(task);
+
+
+    ucc_knomial_pattern_init(team->size, VRANK(team->rank,
+                             coll_task->args.root, team->size),
+                             task->scatter_kn.p.radix, &task->scatter_kn.p);
+    task->scatter_kn.phase = UCC_SCATTER_KN_PHASE_INIT;
+    vroot = ucc_knomial_pattern_loop_rank(p, VRANK(coll_task->args.root,
+                                          coll_task->args.root, team->size));
+    vrank = ucc_knomial_pattern_loop_rank(p, VRANK(team->rank,
+                                          coll_task->args.root, team->size));
+    task->scatter_kn.recv_dist = calc_recv_dist(team->size - p->n_extra, vrank,
+                                                p->radix, vroot);
+    task->scatter_kn.send_offset = 0;
+
+    status = ucc_tl_ucp_scatter_knomial_progress(&task->super);
+    if (UCC_INPROGRESS == status) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+    return ucc_task_complete(coll_task);
+}
+
+ucc_status_t
+ucc_tl_ucp_scatter_knomial_finalize(ucc_coll_task_t *coll_task)
+{
+    return ucc_tl_ucp_coll_finalize(coll_task);
+}
+
+ucc_status_t ucc_tl_ucp_scatter_knomial_init_r(
+    ucc_base_coll_args_t *coll_args, ucc_base_team_t *team,
+    ucc_coll_task_t **task_h, ucc_kn_radix_t radix)
+{
+    ucc_tl_ucp_team_t *tl_team   = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_rank_t         size      = tl_team->size;
+    ucc_rank_t         rank      = tl_team->rank;
+    ucc_tl_ucp_task_t *task;
+
+    task                 = ucc_tl_ucp_init_task(coll_args, team);
+    task->super.post     = ucc_tl_ucp_scatter_knomial_start;
+    task->super.progress = ucc_tl_ucp_scatter_knomial_progress;
+    task->super.finalize = ucc_tl_ucp_scatter_knomial_finalize;
+
+    ucc_assert(coll_args->args.src.info.mem_type ==
+               coll_args->args.dst.info.mem_type);
+    ucc_knomial_pattern_init(size, rank, radix, &task->scatter_kn.p);
+
+    /* In place currently not supported */
+    if (UCC_IS_INPLACE(coll_args->args)) {
+    	return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    *task_h = &task->super;
+    return UCC_OK;
+}
+
+ucc_status_t
+ucc_tl_ucp_scatter_knomial_init(ucc_base_coll_args_t *coll_args,
+                                       ucc_base_team_t *     team,
+                                       ucc_coll_task_t **    task_h)
+{
+    ucc_tl_ucp_team_t *tl_team = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_rank_t         size    = tl_team->size;
+    size_t             count   = coll_args->args.src.info.count;
+    ucc_kn_radix_t     radix;
+
+    radix = ucc_min(UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.bcast_kn_radix, size);
+    radix = 2;
```

Comment:
left from debug?

Reply:
Yes, but was unused because ucc_tl_ucp_bcast_sag_knomial_init() calls directly to ucc_tl_ucp_scatter_knomial_init_r().
Removed.